### ISSUE TITLE
Potential code fix to StoppingCriteria during generation

### ISSUE THREAD

**Issue by mshen2** [allenai/open-instruct#22]
I noticed that for the current StoppingCriteria during text generation is:
```
def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
      sequences_should_be_stopped = []
      for i in range(input_ids.shape[0]):
          for stop_sequence in self.stop_sequences:
              if input_ids[i][-len(stop_sequence):].tolist() == stop_sequence:
                  sequences_should_be_stopped.append(True)
                  break
          sequences_should_be_stopped.append(False)
      return all(sequences_should_be_stopped)
```
However, this will make `all(sequences_should_be_stopped)` always `False` since it does not break out of the outer loop, so model will keep generating even "\n" is already generated (I noticed "\n" is the stopping tokens used in the repo). Isn't something like below should be the correct way? (I know that "\n" is checked afterwards with generated ids, so results won't be affected)
```
def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
      batch_generation_should_be_stopped = []
      for i in range(input_ids.shape[0]):
          sequence_should_be_stopped = False
          for stop_sequence in self.stop_sequences:
              if input_ids[i][-len(stop_sequence):].tolist() == stop_sequence:
                  sequence_should_be_stopped = True
                  break
          batch_generation_should_be_stopped.append(sequence_should_be_stopped)
      return all(batch_generation_should_be_stopped)
```
Another thing I noticed is that since StoppingCriteria happens at batch level, so only all sequences in a batch happens to generate "\n" at a time step will make `__call__`  to return True and generation to halt, which is rare, so model will most likely to keep generating until hitting max length to generate or hitting model eos token in most batches. So I am wondering isn't using StoppingCriteria plus batch size of 1 more efficient during generation based inference?

**Comment by yizhongw** [comment]
Hi @mshen2, thanks for pointing out this! I have fixed the issue in PR #27 .

**Comment by yizhongw** [comment]
I also agree with using a larger batch size might slow down the generation, but it depends on how different the generation lengths are for the instances in a batch.
